### PR TITLE
SONAR-7330 When searching active rules by profile, ignore active rule on removed rules

### PR DESCRIPTION
**Pull request by julienlancelot**
SONAR-7330 When searching active rules by profile, ignore active rule on removed rules

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java [hunk 1]
```diff
@@ -213,4 +223,22 @@ private static RuleActivation toRuleActivation(org.sonar.api.rules.ActiveRule ac
...
+    public Integer apply(@Nonnull ActiveRuleDto input) {
+      return input.getId();
+    }
+  }
+
+  private enum ActiveRuleParamDtoToActiveRuleId implements Function<ActiveRuleParamDto, Integer> {
+    INSTANCE;
+
+    @Override
+    public Integer apply(@Nonnull ActiveRuleParamDto input) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java]: MAJOR The "input" parameter nullability is different in the superclass method, and that should not be changed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java [hunk 2]
```diff
@@ -124,10 +129,15 @@ private RulesProfile wrap(QualityProfileDto profile) {
     DbSession dbSession = dbClient.openSession(false);
     RulesProfile target = new RulesProfile(profile.getName(), profile.getLanguage());
     try {
-      for (ActiveRuleDto activeRule : dbClient.activeRuleDao().selectByProfileKey(dbSession, profile.getKey())) {
+      List<ActiveRuleDto> activeRuleDtos = dbClient.activeRuleDao().selectByProfileKey(dbSession, profile.getKey());
+      List<ActiveRuleParamDto> activeRuleParamDtos = dbClient.activeRuleDao().selectParamsByActiveRuleIds(dbSession, Lists.transform(activeRuleDtos, ActiveRuleDtoToId.INSTANCE));
+      ListMultimap<Integer, ActiveRuleParamDto> activeRuleParamsByActiveRuleId = FluentIterable.from(activeRuleParamDtos).index(ActiveRuleParamDtoToActiveRuleId.INSTANCE);
+
+      for (ActiveRuleDto activeRule : activeRuleDtos) {
+        // TODO all rules should be loaded by using one query with all active rule keys as parameter
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java]: INFO Complete the task associated to this TODO comment. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java [hunk 3]
```diff
@@ -213,4 +223,22 @@ private static RuleActivation toRuleActivation(org.sonar.api.rules.ActiveRule ac
     return ruleActivation;
   }
 
+  private enum ActiveRuleDtoToId implements Function<ActiveRuleDto, Integer> {
+    INSTANCE;
+
+    @Override
+    public Integer apply(@Nonnull ActiveRuleDto input) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java]: MAJOR The "input" parameter nullability is different in the superclass method, and that should not be changed. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2638)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java [hunk 4]
```diff
@@ -124,10 +129,15 @@ private RulesProfile wrap(QualityProfileDto profile) {
     DbSession dbSession = dbClient.openSession(false);
     RulesProfile target = new RulesProfile(profile.getName(), profile.getLanguage());
     try {
-      for (ActiveRuleDto activeRule : dbClient.activeRuleDao().selectByProfileKey(dbSession, profile.getKey())) {
+      List<ActiveRuleDto> activeRuleDtos = dbClient.activeRuleDao().selectByProfileKey(dbSession, profile.getKey());
+      List<ActiveRuleParamDto> activeRuleParamDtos = dbClient.activeRuleDao().selectParamsByActiveRuleIds(dbSession, Lists.transform(activeRuleDtos, ActiveRuleDtoToId.INSTANCE));
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java]: I think we should create `ActiveRuleDtoFunctions` similar to `RuleDtoFunctions` and `ComponentDtoFunctions` with a method `toId()`.

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java [hunk 5]
```diff
@@ -124,10 +129,15 @@ private RulesProfile wrap(QualityProfileDto profile) {
...
+      ListMultimap<Integer, ActiveRuleParamDto> activeRuleParamsByActiveRuleId = FluentIterable.from(activeRuleParamDtos).index(ActiveRuleParamDtoToActiveRuleId.INSTANCE);
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java]: I think we should create `ActiveRuleParamDtoFunctions` similar to `RuleDtoFunctions` with a method `toRuleId()`.

File: server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java [hunk 6]
```diff
@@ -100,9 +104,11 @@ public void start() {
         }
       }
       List<RuleDto> activeRules = processRemainingDbRules(allRules.values(), session);
-      removeActiveRulesOnStillExistingRepositories(session, activeRules, context);
+      List<ActiveRuleChange> changes = removeActiveRulesOnStillExistingRepositories(session, activeRules, context);
       session.commit();
       ruleIndexer.setEnabled(true).index();
+      activeRuleIndexer.setEnabled(true);
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java]: shouldn't it be RuleActivator responsability to update the index?

File: server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java (hunk 5 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileExporters.java]: You're right !

File: server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java (hunk 6 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java]: How ?
Indexing can only be done after the commit, and a lot of things can be done before commiting

**Comment by julienlancelot** [comment]
@teryk Feedback applied

**Pull request merged**